File: Capgemini/423.reconstruct-original-digits-from-english.java
/**
 * Algorithm:
 * Count letters, then recover digits with unique identifying letters:
 * z->0, w->2, u->4, x->6, g->8. The remaining digit counts are derived after
 * subtracting those already fixed digits.
 *
 * Java data structures:
 * Fixed int arrays replace Python Counter. The alphabet and digit set are
 * constant size, so access is O(1).
 *
 * Complexity:
 * Time O(n), auxiliary space O(1).
 */
class Solution {
    public String originalDigits(String s) {
        int[] count = new int[26];
        for (int i = 0; i < s.length(); i++) {
            count[s.charAt(i) - 'a']++;
        }

        int[] digit = new int[10];
        digit[0] = count['z' - 'a'];
        digit[2] = count['w' - 'a'];
        digit[4] = count['u' - 'a'];
        digit[6] = count['x' - 'a'];
        digit[8] = count['g' - 'a'];
        digit[3] = count['h' - 'a'] - digit[8];
        digit[5] = count['f' - 'a'] - digit[4];
        digit[7] = count['s' - 'a'] - digit[6];
        digit[1] = count['o' - 'a'] - digit[0] - digit[2] - digit[4];
        digit[9] = count['i' - 'a'] - digit[5] - digit[6] - digit[8];

        StringBuilder ans = new StringBuilder();
        for (int i = 0; i < 10; i++) {
            for (int c = 0; c < digit[i]; c++) {
                ans.append(i);
            }
        }
        return ans.toString();
    }
}

File: JPMorgan Chase/423.reconstruct-original-digits-from-english.cpp
#include <algorithm>
#include <array>
#include <cmath>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <random>
#include <regex>
#include <set>
#include <sstream>
#include <string>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;

class Solution {
public:
    string originalDigits(string s) {
        vector<int> cnt(26), digit(10);
        for (char c : s) ++cnt[c - 'a'];
        digit[0] = cnt['z' - 'a'];
        digit[2] = cnt['w' - 'a'];
        digit[4] = cnt['u' - 'a'];
        digit[6] = cnt['x' - 'a'];
        digit[8] = cnt['g' - 'a'];
        digit[3] = cnt['h' - 'a'] - digit[8];
        digit[5] = cnt['f' - 'a'] - digit[4];
        digit[7] = cnt['s' - 'a'] - digit[6];
        digit[1] = cnt['o' - 'a'] - digit[0] - digit[2] - digit[4];
        digit[9] = cnt['i' - 'a'] - digit[5] - digit[6] - digit[8];
        string ans;
        for (int d = 0; d <= 9; ++d) ans.append(digit[d], char('0' + d));
        return ans;
    }
};

/*
Interview explanation:
Unique letters identify several digits first: z, w, u, x, g correspond to 0,2,4,6,8. After those counts are known, remaining digits are determined by letters whose ambiguity has been removed.

C++ data structures: vector<int> of size 26 is faster and simpler than unordered_map for fixed lowercase letters.

Edge cases: repeated digits are emitted by string::append(count, char), and the loop emits them sorted.

Complexity: O(n) time and O(1) space.
*/
